android-voice-processor/src/main/java/ai/picovoice/android/voiceprocessor/VoiceProcessor.java

```diff
@@ -0,0 +1,305 @@
+     * {@link #addBufferListener(VoiceProcessorBufferListener)} in order to receive audio
+     * frames from the VoiceProcessor.
+     */
+    public synchronized void start() {
+        if (getIsRecording()) {
+            return;
+        }
+
+        readThread = Executors.newSingleThreadExecutor().submit(new Callable<Void>() {
+            @Override
+            public Void call() {
+                android.os.Process.setThreadPriority(Process.THREAD_PRIORITY_URGENT_AUDIO);
+                read();
+                return null;
+            }
+        });
+    }
+
+    /**
+     * Stops audio capture. Frames will stop being delivered to the subscribed listeners.
+     *
+     * @throws VoiceProcessorException if an error is encountered while trying to stop the
+     *                                 recorder thread.
+     */
+    public synchronized void stop() throws VoiceProcessorException {
+        if (!getIsRecording()) {
+            return;
+        }
+
+        isStopRequested.set(true);
+        try {
+            readThread.get();
+            readThread = null;
+        } catch (ExecutionException | InterruptedException e) {
+            throw new VoiceProcessorException(
+                    "An error was encountered while requesting to stop the audio recording",
+                    e);
+        } finally {
+            isStopRequested.set(false);
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private void read() {
+        final int minBufferSize = AudioRecord.getMinBufferSize(
+                this.sampleRate,
+                AudioFormat.CHANNEL_IN_MONO,
+                AudioFormat.ENCODING_PCM_16BIT);
+        final int bufferSize = Math.max(this.sampleRate / 2, minBufferSize);
+
+        AudioRecord recorder;
+        try {
+            recorder = new AudioRecord(
+                    MediaRecorder.AudioSource.MIC,
+                    this.sampleRate,
+                    AudioFormat.CHANNEL_IN_MONO,
+                    AudioFormat.ENCODING_PCM_16BIT,
+                    bufferSize);
+        } catch (IllegalArgumentException e) {
+            onError(new VoiceProcessorException(
+                    "Unable to initialize audio recorder with required parameters",
+                    e));
+            return;
+        }
+
+        if (recorder.getState() != AudioRecord.STATE_INITIALIZED) {
+            onError(new VoiceProcessorException(
+                    "Audio recorder did not initialize successfully. " +
+                            "Ensure you have acquired permission to record audio from the user."));
+            return;
+        }
+
+        final short[] buffer = new short[this.frameLength];
+        try {
+            recorder.startRecording();
+
+            while (!isStopRequested.get()) {
+                if (recorder.read(buffer, 0, buffer.length) == buffer.length) {
+                    onBuffer(buffer);
+                }
```

Comment:
I went with putting an event into the `errorListener` because it's also possible that the recorder isn't giving any bytes or more than the frame. In that case they need to correct their code.